src/util/fifo.h

```diff
@@ -1,69 +1,116 @@
+        count = std::min(available, count);
+        *sizePtr1 = std::min(m_size - (writeIndex & m_mask), count);
+        *sizePtr2 = count - *sizePtr1;
+        *dataPtr1 = m_data.data() + (writeIndex & m_mask);
+        *dataPtr2 = *sizePtr2 == 0 ? nullptr : m_data.data();
+        return count;
     }
-    int releaseWriteRegions(int count) {
-        return PaUtil_AdvanceRingBufferWriteIndex(&m_ringBuffer, count);
+    uint32_t releaseWriteRegions(uint32_t count) {
+        const uint32_t writeIndex = (m_writeIndex.load(std::memory_order_relaxed) + count);
+        m_writeIndex.store(writeIndex, std::memory_order_release);
+        return (writeIndex & m_mask);
     }
-    int aquireReadRegions(int count,
-            DataType** dataPtr1, ring_buffer_size_t* sizePtr1,
-            DataType** dataPtr2, ring_buffer_size_t* sizePtr2) {
-        return PaUtil_GetRingBufferReadRegions(&m_ringBuffer, count,
-                (void**)dataPtr1, sizePtr1, (void**)dataPtr2, sizePtr2);
+    uint32_t aquireReadRegions(uint32_t count,
+            DataType** dataPtr1,
+            uint32_t* sizePtr1,
+            DataType** dataPtr2,
+            uint32_t* sizePtr2) {
+        const uint32_t readIndex = m_readIndex.load(std::memory_order_relaxed);
+        const uint32_t writeIndex = m_writeIndex.load(std::memory_order_acquire);
+        const uint32_t available = writeIndex - readIndex;
+        count = std::min(available, count);
+        *sizePtr1 = std::min(m_size - (readIndex & m_mask), count);
+        *sizePtr2 = count - *sizePtr1;
+        *dataPtr1 = m_data.data() + (readIndex & m_mask);
+        *dataPtr2 = *sizePtr2 == 0 ? nullptr : m_data.data();
+
+        return count;
     }
-    int releaseReadRegions(int count) {
-        return PaUtil_AdvanceRingBufferReadIndex(&m_ringBuffer, count);
+    uint32_t releaseReadRegions(uint32_t count) {
+        const uint32_t readIndex = (m_readIndex.load(std::memory_order_relaxed) + count);
+        m_readIndex.store(readIndex, std::memory_order_release);
+        return (readIndex & m_mask);
     }
-    int flushReadData(int count) {
-        int flush = math_min(readAvailable(), count);
-        return PaUtil_AdvanceRingBufferReadIndex(&m_ringBuffer, flush);
+    uint32_t flushReadData(uint32_t count) {
+        uint32_t readIndex = m_readIndex.load(std::memory_order_relaxed);
+        const uint32_t writeIndex = m_writeIndex.load(std::memory_order_acquire);
+        const uint32_t available = writeIndex - readIndex;
+        count = std::min(available, count);
+        readIndex += count;
+        m_readIndex.store(readIndex, std::memory_order_release);
+        return (readIndex & m_mask);
     }
 
   private:
+    const uint32_t m_size;
+    const uint32_t m_mask;
     std::vector<DataType> m_data;
-    PaUtilRingBuffer m_ringBuffer;
+    std::atomic<uint32_t> m_writeIndex;
+    std::atomic<uint32_t> m_readIndex;
     DISALLOW_COPY_AND_ASSIGN(FIFO);
```

Comment:
can you implement move semantics and adhere to the rule of 5?

Reply:
I’m just following the original FIFO here, apparently this was sufficient for the current use.

Reply:
Haven't looked at the diff, but I have a suspicion that the original implementation was written before move semantics were a thing (`C++11`), so we might as well go the extra step. We then may also look for unnecessarily heap-allocated `FIFO`s that workaround the lack of move semantics and inline those (the eliminated pointer dereference may result in a nice runtime speedup).

Reply:
Ok, but let’s do that in a follow up PR. The goal of these PRs is to get rid of all the tsan warnings.

Reply:
The heap allocation happens twice. One time by the std::vector and probably a second time for the control structure. 
We may consider to use a std::array() based version where the size becomes a template parameter. 
This way we get also around of the unneccessary default initalization of the vector.

Reply:
If its feasible to shift the size calculation to compile-time, I'm all for it.

Reply:
I haven't looked at the actual use of this, so I don't know if using size as a template argument is even an option. Anyway, let's keep additional improvements beyond fixing the tsan issue for later PRs.